### PR TITLE
Fix: generate_series function support string type

### LINKED ISSUE #11922: Internal error `generate_series`: Internal error: could not cast value to .....arrow_array::types::Date32Type`
### Describe the bug

When running a query like 

```sql
select generate_series(start, '1993-03-01', INTERVAL '1 year') from date_table;
```

You get an error:

> DataFusion error: Internal error: could not cast value to arrow_array::array::primitive_array::PrimitiveArray<arrow_array::types::Date32Type>.
> This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker

### To Reproduce

```sql
DataFusion CLI v40.0.0
> CREATE TABLE date_table(
  start DATE,
  stop DATE,
  step INTERVAL
) AS VALUES
  (DATE '1992-01-01', DATE '1993-01-02', INTERVAL '1' MONTH),
  (DATE '1993-02-01', DATE '1993-01-01', INTERVAL '-1' DAY),
  (DATE '1989-04-01', DATE '1993-03-01', INTERVAL '1' YEAR);
0 row(s) fetched.
Elapsed 0.016 seconds.

> select generate_series(start, '1993-03-01', INTERVAL '1 year') from date_table;

Internal error: could not cast value to arrow_array::array::primitive_array::PrimitiveArray<arrow_array::types::Date32Type>.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### Expected behavior

I expect you get the same answer as if you explicitly cast the second argument to `date`:

```sql
> select generate_series(start, '1993-03-01'::DATE, INTERVAL '1 year') from date_table;
+-------------------------------------------------------------------------------------------------------------------------------------------+
| generate_series(date_table.start,Utf8("1993-03-01"),IntervalMonthDayNano("IntervalMonthDayNano { months: 12, days: 0, nanoseconds: 0 }")) |
+-------------------------------------------------------------------------------------------------------------------------------------------+
| [1992-01-01, 1993-01-01]                                                                                                                  |
| [1993-02-01]                                                                                                                              |
| [1989-04-01, 1990-04-01, 1991-04-01, 1992-04-01]                                                                                          |
+-------------------------------------------------------------------------------------------------------------------------------------------+
3 row(s) fetched.
Elapsed 0.001 seconds.
```

### Additional context

I added a test here: https://github.com/apache/datafusion/pull/11921

### PR DESCRIPTION
**Pull request by getChan**
## Which issue does this PR close?

Closes #11922.

## Rationale for this change
`generate_series` function does not accept string arguments. 
However, the as-is test provide string  arguments.

I think it would be better if the type signature of `generate_series` was changed.
https://github.com/apache/datafusion/blob/main/datafusion/functions-nested/src/range.rs#L129

## What changes are included in this PR?
Only sqllogictest

## Are these changes tested?
Yes.

## Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by jayzhan211** [comment]
@getChan I think what we need is to support string type but not modifying the test to other already supported type


```
D create table date_table(start date, stop date, step interval);
D insert into date_table values (DATE '1992-01-01', DATE '1993-01-02', INTERVAL '1' MONTH);
D select generate_series(start, '1993-03-01', INTERVAL '1 year') from date_table;
┌────────────────────────────────────────────────────────────────────┐
│ generate_series("start", '1993-03-01', CAST('1 year' AS INTERVAL)) │
│                            timestamp[]                             │
├────────────────────────────────────────────────────────────────────┤
│ [1992-01-01 00:00:00, 1993-01-01 00:00:00]                         │
└────────────────────────────────────────────────────────────────────┘
```

We should able to return the same result as Duckdb's

**Comment by getChan** [comment]
In that case, I guess we should extend the above issue to "Add string type support to generate_series function".

**Comment by jayzhan211** [comment]
> I expect you get the same answer as if you explicitly cast the second argument to date:

The goal is make the query return the same result the same as the one you explicitly cast the argument.
Supporting string type is one of the possible solution for this.

>  In that case, I guess we should extend the above issue to "Add string type support to generate_series function".
Yes, I think what we need is to support string type

**Comment by alamb** [comment]
> Yes, I think what we need is to support string type, or cast the string to date. Not sure what is the exact solution. But I expect we should get the correct result for that query

The type coercsion system is supposed to handle adding this kind of cast. 

I wonder if we can could remove the `Any(3)` signature  a new signature to `generate_series` :

https://github.com/apache/datafusion/blob/e24a5dd5a508ba605686206c853c914a2d0cc095/datafusion/functions-nested/src/range.rs#L125-L129

I think that would then ensure DataFusion actually casts the arguments to the supported types

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

**Comment by getChan** [comment]
> I wonder if we can could remove the Any(3) signature a new signature to generate_series :

removed `Any(3)` signature. and fix some type coerce code for `range`, `generate_series`

File: datafusion/functions-nested/src/range.rs [hunk 1]
```diff
@@ -317,6 +315,14 @@ fn gen_range_date(args: &[ArrayRef], include_upper: bool) -> Result<ArrayRef> {
     let values_builder = Date32Builder::new();
     let mut list_builder = ListBuilder::new(values_builder);
 
+    if start_array.unwrap().is_null(0)
```
Comment (jayzhan211) [inline datafusion/functions-nested/src/range.rs:318]: Actually this function's signature is quite complex, I think we should use `UserDefined` signature to handle all possible types and **coerce** them to the expected one with `coerce_types`. Then we don't need to handle null inside the actual computation here

File: datafusion/expr/src/type_coercion/functions.rs [hunk 2]
```diff
@@ -580,7 +583,7 @@ fn coerced_from<'a>(
         {
             Some(type_into.clone())
         }
-        (Interval(_), _) if matches!(type_from, Utf8 | LargeUtf8) => {
+        (Interval(_), _) if matches!(type_from, Utf8 | LargeUtf8 | Null) => {
```
Comment (jayzhan211) [inline datafusion/expr/src/type_coercion/functions.rs:586]: This is what I want to deprecate but not yet, at least we don't make it more complex

File: datafusion/functions-nested/src/range.rs (hunk 1 above)
Comment (getChan) [reply datafusion/functions-nested/src/range.rs]: thanks for review. 
replace type coerce with implement coerce_types method.

File: datafusion/expr/src/type_coercion/functions.rs (hunk 2 above)
Comment (getChan) [reply datafusion/expr/src/type_coercion/functions.rs]: rollback codes in `functions.rs`

File: datafusion/sqllogictest/test_files/array.slt [hunk 3]
```diff
@@ -5804,7 +5804,7 @@ select generate_series(5),
 ----
 [0, 1, 2, 3, 4, 5] [2, 3, 4, 5] [2, 5, 8] [1, 2, 3, 4, 5] [5, 4, 3, 2, 1] [10, 7, 4] [1992-09-01, 1992-10-01, 1992-11-01, 1992-12-01, 1993-01-01, 1993-02-01, 1993-03-01] [1993-02-01, 1993-01-31, 1993-01-30, 1993-01-29, 1993-01-28, 1993-01-27, 1993-01-26, 1993-01-25, 1993-01-24, 1993-01-23, 1993-01-22, 1993-01-21, 1993-01-20, 1993-01-19, 1993-01-18, 1993-01-17, 1993-01-16, 1993-01-15, 1993-01-14, 1993-01-13, 1993-01-12, 1993-01-11, 1993-01-10, 1993-01-09, 1993-01-08, 1993-01-07, 1993-01-06, 1993-01-05, 1993-01-04, 1993-01-03, 1993-01-02, 1993-01-01] [1989-04-01, 1990-04-01, 1991-04-01, 1992-04-01]
 
-query error DataFusion error: Execution error: unsupported type for range. Expected Int64 or Date32, got: Timestamp\(Nanosecond, None\)
+query error DataFusion error: Execution error: Cannot generate date range less than 1 day\.
```
Comment (getChan) [inline datafusion/sqllogictest/test_files/array.slt:5807]: is it okay for the function to support the timestamp type?

Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/array.slt:5807]: We usually follow what Postgres or DuckDB does

File: datafusion/functions-nested/src/range.rs [hunk 4]
```diff
@@ -75,9 +65,34 @@ impl ScalarUDFImpl for Range {
         &self.signature
     }
 
+    fn coerce_types(&self, _arg_types: &[DataType]) -> Result<Vec<DataType>> {
+        _arg_types
```
Comment (jayzhan211) [inline datafusion/functions-nested/src/range.rs:69]: nit: `arg_types`

File: datafusion/sqllogictest/test_files/array.slt (hunk 3 above)
Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/array.slt:5807]:
```
D select generate_series('2021-01-01'::timestamp, '2021-01-02'::timestamp, INTERVAL '1' HOUR);
┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│   generate_series(CAST('2021-01-01' AS TIMESTAMP), CAST('2021-01-02' AS TIMESTAMP), to_hours(CAST(trunc(CAST('1' AS DOUBLE)) AS BIGINT)))   │
│                                                                 timestamp[]                                                                 │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ [2021-01-01 00:00:00, 2021-01-01 01:00:00, 2021-01-01 02:00:00, 2021-01-01 03:00:00, 2021-01-01 04:00:00, 2021-01-01 05:00:00, 2021-01-01…  │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

**Review by jayzhan211 — APPROVED** [review]
👍

**Review by jayzhan211 — APPROVED** [review]
Thanks @getChan

File: datafusion/sqllogictest/test_files/array.slt (hunk 3 above)
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/array.slt:5807]: I will file an ticket for this issue

**Review by alamb — COMMENTED** [review]
Thank you 🙏

**Pull request merged**